src/core/services-impl/src/main/java/it/geosolutions/geostore/services/ResourceServiceImpl.java

```diff
@@ -491,72 +493,82 @@ public long getCountByFilter(SearchFilter filter)
-                                authUserIsOwner = true;
-                                if (rule.isCanWrite()) {
-                                    shortResource.setCanEdit(true);
-                                    shortResource.setCanDelete(true);
-
-                                    break;
-                                }
-                            }
-                        } else if (userGroup != null) {
-                            List<String> groups = extractGroupNames(authUser.getGroups());
-                            if (groups.contains(userGroup.getGroupName())) {
-                                if (rule.isCanWrite()) {
-                                    shortResource.setCanEdit(true);
-                                    shortResource.setCanDelete(true);
-
-                                    break;
-                                }
-                            }
-                        }
-                    }
-                    if (!authUserIsOwner) resourceCanBeListed = resource.isAdvertised();
+    private List<ShortResource> convertToShortList(List<Resource> resources, User user) {
+        return resources.stream()
+                .filter(r -> isResourceCanBeListed(user, r))
+                .map(r -> createShortResourceWithPermissions(user, r))
+                .collect(Collectors.toList());
+    }
+
+    private boolean isResourceCanBeListed(User user, Resource resource) {
+        if (user != null
+                && (user.getRole().equals(Role.ADMIN) || isUserResourceOwner(user, resource))) {
+            return true;
+        }
+        return resource.isAdvertised();
+    }
+
+    private boolean isUserResourceOwner(User user, Resource resource) {
+        for (SecurityRule securityRule : resource.getSecurity()) {
+            User owner = securityRule.getUser();
+            if (owner != null) {
+                if (owner.getId().equals(user.getId())) {
+                    return true;
                 }
-            } else {
-                resourceCanBeListed = resource.isAdvertised();
             }
+        }
+        return false;
+    }
+
+    private ShortResource createShortResourceWithPermissions(User user, Resource resource) {
+        ShortResource shortResource = new ShortResource(resource);
 
-            if (resourceCanBeListed) swList.add(shortResource);
+        if (user != null) {
+            addPermissionsToShortResource(user, resource, shortResource);
         }
 
-        return swList;
+        return shortResource;
     }
 
-    public static List<String> extractGroupNames(Set<UserGroup> groups) {
-        List<String> groupNames = new ArrayList<String>();
-        if (groups == null) {
-            return groupNames;
+    /**
+     * Checks if the user can modify and delete the loaded resource (and associated attributes and
+     * stored data). This to inform the client in HTTP response result.
+     */
+    public ShortResource addPermissionsToShortResource(
```

Comment:
Possible Improvement:
Unify permission logic in a single service or utility. E.g., a PermissionService that says hasWritePermission(User, Resource). This ensures you don’t accidentally diverge logic in multiple places.

Reply:
Created a specific `PermissionService` to handle resource access checks.